### PR TITLE
[Feat] Add official LTX-2.3 multimodal guidance and align inference numerics

### PR DESCRIPTION
**Pull request by mglyn**
## Purpose

This PR updates LTX-2.3 T2V/I2V to follow the official one-stage non-HQ recipe, exposes the missing multimodal guidance controls, and moves LTX-2.3 into an isolated `ltx2_3` package so these precision changes do not affect LTX-2.

Main changes:

- Add video/audio CFG, STG, modality guidance, rescale, and STG block controls.
- Align numerically visible behavior with the official pipeline: connector attention, packed latent sampling, RoPE dtype, x0/noise conversion, Euler update order, and decode conditioning.

Implementation layout:

- `ltx2_3_pipeline_base.py`: lifecycle and stage execution
- `ltx2_3_guidance.py`: guidance math and pass metadata
- `ltx2_3_guidance_parallel.py`: CFG-parallel strategy
- `ltx2_3_latent_preparation.py`: latent pack/unpack/normalize/decode conditioning
- `ltx2_3_denoise_scheduling.py`: official sigma schedule and video/audio scheduler
- `ltx2_3_connector.py`: official connector attention override
- `ltx2_3_recipes.py`: official one-stage, two-stage, and HQ recipe metadata

## Test Plan

- Unit tests for LTX-2.3 request parsing, guidance defaults, T2V/I2V behavior, CFG-parallel compatibility, latent helpers, and model-extra registration.
- LTX-2 regression coverage for the package split.
- performance / quality sweep.

**vLLM Version:** 0.24.0

**vLLM-Omni Commit:** f64dce30690a264868c7f22889b7298a9ce440b9

## Test Result

<details>
<summary>unit tests</summary>

Focused LTX-2.3 unit tests:

```bash
/root/vllm-omni/.venv/bin/python -m pytest tests/diffusion/models/ltx2_3 -q
```

```text
66 passed, 21 warnings
```

LTX-2 regression tests:

```bash
/root/vllm-omni/.venv/bin/python -m pytest tests/diffusion/models/ltx2 -q
```

```text
5 passed, 21 warnings
```

</details>

<details>
<summary>precision conclusions</summary>

Precision probe against the official pipeline with `DIFFUSION_ATTENTION_BACKEND=CUDNN_ATTN`:

- Denoise after-step video/audio latents are bitwise equal for all 20 steps.
- Final packed and unpacked latents are bitwise equal.
- Raw VAE decode is nearly identical: RGB SSIM `0.99958`, PSNR `60.11 dB`.
- Pre-encode uint8 RGB frames: mean SSIM `0.99758` over `1024x576`, `81f`.
- MP4 e2e SSIM is lower because official and Omni use different video writing paths; official raw output vs official MP4 is already about `0.961` RGB SSIM.

Default `FLASH_ATTN` precision smoke:

- L20 local official-GT keyframes pass with SSIM/PSNR: frame 0 `0.993917` / `43.97 dB`, frame 60 `0.987650` / `40.13 dB`, frame 120 `0.977313` / `37.99 dB`.
- This is intentionally reported separately from the `CUDNN_ATTN` raw-tensor alignment above; `FLASH_ATTN` is the production default backend, while `CUDNN_ATTN` is the closest official numerical comparison path.

</details>

<details>
<summary>video comparison</summary>

Official vs Omni one-stage comparison video config:

- Platform: single NVIDIA L20, CPU/layerwise offload, eager execution.
- Prompt: `Floating crystal islands in cosmic starry sky, glowing nebula, soft luminous particles flowing around, slow camera rotation, dreamlike atmosphere, ultra-detailed, surreal fantasy scene`.
- Negative prompt: official LTX-2.3 default negative prompt.
- Shape: `512x384`, `161` frames, `24` fps, `30` denoise steps, seed `42`.
- Video guidance: CFG `3.0`, STG `1.0`, modality `3.0`, rescale `0.7`, STG blocks `[28]`.
- Audio guidance: CFG `7.0`, STG `1.0`, modality `3.0`, rescale `0.7`, STG blocks `[28]`.

</details>

https://github.com/user-attachments/assets/960bc2a3-97c1-4909-9602-2f0e3ea0e8c5

https://github.com/user-attachments/assets/d49de1d3-9543-44f2-b668-ab460e440f4e

https://github.com/user-attachments/assets/7ac725e0-06ae-4a84-a5f9-37ad34d7d16f

<details>
<summary>feature coverage sweep</summary>

sweep feature coverage validated for this PR:

| Pipeline | torch.compile | ⚡TeaCache | ⚡Cache-DiT | 🔀SP (Ulysses & Ring) | 🔀CFG-Parallel | 🔀Tensor-Parallel | Pipeline-Parallel | 🔀HSDP | 💾CPU Offload (Layerwise) | 💾VAE-Patch-Parallel | 💾Quantization | Request batching | Step Execution |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
| **T2V** | ✅ | ❌ | ✅ | ✅ | ✅ | ✅ | ❌ | ❌ | ✅ | ✅ (decode) | ❌ | ✅ (2 requests) | ❌ |
| **I2V** | ✅ | ❌ | ✅ | ✅ | ✅ | ✅ | ❌ | ❌ | ✅ | ✅ (decode) | ❌ | ✅ (2 requests) | ❌ |

### A100 stage-0 latency

Measured on 4x A100 80GB with official one-stage full guidance, `512x384`, `25f`, and `10` denoise steps. T2V/I2V feature cells use three single requests; `batch2` uses four requests admitted in pairs, so its time is per two-request batch.

| Feature | T2V mean (ms) | I2V mean (ms) | Peak GPU MiB (T2V / I2V) |
|:---|---:|---:|---:|
| Base | 7,033.3 | 7,049.5 | 72,491 / 72,492 |
| torch.compile | 5,938.5 | 6,015.2 | 72,492 / 72,492 |
| Cache-DiT | 6,995.4 | 6,905.1 | 72,737 / 72,739 |
| Ulysses 2 | 5,899.0 | 6,073.4 | 72,492 / 72,492 |
| Ring 2 | 5,970.7 | 5,996.8 | 72,499 / 72,500 |
| TP 2 | 6,726.7 | 6,767.8 | 54,052 / 54,054 |
| VAE patch 2 | 7,022.9 | 7,031.4 | 72,491 / 72,492 |
| Layerwise offload | 15,539.5 | 15,456.3 | 37,812 / 37,812 |
| Batch 2 (per batch) | 14,171.8 | 14,238.4 | 76,018 / 76,019 |

</details>

### CODE REVIEW

**Comment by TaffyOfficial** [comment]
I think there is a batching bug in the LTX2.3 guidance parsing. _resolve_request_inputs() uses the first request’s guidance_scale and then resolves all later requests with that same value. So if two requests in the same batch explicitly set different guidance_scale values, the later one can be silently parsed with the first request’s value instead of being rejected. Since this code already requires homogeneous guidance params within a batch, we should resolve each request with its own explicit guidance_scale and add a unit test for the mixed-guidance-scale case.
Also, the recipe abstraction looks only partially wired. The data model has separate video_guidance and audio_guidance, but _resolve_guidance_params() still uses recipe-level shortcuts like recipe.rescale_scale, which currently read from video guidance. That is fine only because the default one-stage recipe mostly has matching video/audio values, but the HQ recipe already has different video/audio rescale defaults. Either the parsing should use recipe.video_guidance / recipe.audio_guidance directly, or the unused HQ/two-stage recipes should not be exposed yet.

**Comment by mglyn** [comment]
Thanks, both points were valid.
I updated _resolve_request_inputs() so each request resolves guidance params with its own explicit guidance_scale, and added a mixed-guidance-scale batch unit test.
I also removed the recipe-level video-only guidance shortcuts from parsing. _resolve_guidance_params() now reads recipe.video_guidance and recipe.audio_guidance directly, including split audio/video rescale defaults. @TaffyOfficial 
Related tracking issue #4983 #4985.

File: .buildkite/test-nightly.yml [hunk 1]
```diff
@@ -1440,7 +1440,6 @@ steps:
         commands:
           - pytest -s -v tests/e2e/accuracy/hunyuanvideo15_t2v/test_hunyuanvideo15_t2v_video_similarity.py -m full_model --run-level full_model
           - pytest -s -v tests/e2e/accuracy/hunyuanvideo15_i2v/test_hunyuanvideo15_i2v_video_similarity.py -m full_model --run-level full_model
-          - pytest -s -v tests/e2e/accuracy/test_ltx2_3_video_similarity.py -m full_model --run-level full_model
```
Comment (david6666666) [inline .buildkite/test-nightly.yml:1443]: Can we update this accuracy test for the new recipe instead of dropping the only LTX-2.3 E2E regression guard?

**Comment by mglyn** [comment]
This PR is too large, it won’t continue moving forward. I’ll provide the implementation in smaller, split-out PRs instead.

File: vllm_omni/diffusion/models/ltx2_3/ltx2_3_guidance.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+) -> torch.Tensor:
+    dtype = cond.dtype
+    cond = cond.float()
+    uncond_text = uncond_text.float() if isinstance(uncond_text, torch.Tensor) else uncond_text
+    uncond_perturbed = uncond_perturbed.float() if isinstance(uncond_perturbed, torch.Tensor) else uncond_perturbed
+    uncond_modality = uncond_modality.float() if isinstance(uncond_modality, torch.Tensor) else uncond_modality
+    pred = (
+        cond
+        + (cfg_scale - 1) * (cond - uncond_text)
+        + stg_scale * (cond - uncond_perturbed)
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/ltx2_3/ltx2_3_guidance.py:101]: `do_stg` gates the ptb pass on `stg_blocks` too, but this term still applies `stg_scale` with `uncond_perturbed` defaulting to `0.0` when the pass is skipped. So `stg_scale=1.0` + `stg_blocks=[]` silently adds a spurious `+ cond`. Force `stg_scale` to 0 in `_resolve_guidance_params` when the blocks are empty.

File: vllm_omni/diffusion/models/ltx2_3/ltx2_3_guidance.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    sigma = sigmas[step_index].to(torch.float32)
+    sigma_next = sigmas[step_index + 1].to(torch.float32)
+    dt = sigma_next - sigma
+    return (sample.to(torch.float32) + velocity.to(torch.float32) * dt).to(sample.dtype)
+
+
+def rescale_guided_x0(cond: torch.Tensor, pred: torch.Tensor, rescale_scale: float) -> torch.Tensor:
+    if rescale_scale == 0.0:
+        return pred
+    factor = cond.float().std() / pred.float().std()
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/ltx2_3/ltx2_3_guidance.py:77]: `cond.float().std()` / `pred.float().std()` collapse the whole batch into one scalar, so `num_videos_per_prompt>1` mixes samples. Reduce over the non-batch dims (`dim=(1,2), keepdim=True`) so it rescales per-sample. Also no guard if `pred.std()==0`.

File: vllm_omni/diffusion/models/ltx2_3/ltx2_3_denoise_scheduling.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        tokens = max_anchor
+
+        sigmas = torch.linspace(1.0, 0.0, num_inference_steps + 1)
+        slope = (max_shift - base_shift) / (max_anchor - base_anchor)
+        intercept = base_shift - slope * base_anchor
+        sigma_shift = tokens * slope + intercept
+        exp_shift = math.exp(sigma_shift)
+        sigmas = torch.where(sigmas != 0, exp_shift / (exp_shift + (1 / sigmas - 1)), 0)
+
+        terminal = config.get("shift_terminal", 0.1)
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/ltx2_3/ltx2_3_denoise_scheduling.py:119]: The `0.1` fallback always stretches to a terminal even when the scheduler config has no `shift_terminal`, but upstream treats a missing value as a no-op. Default to `None` and skip the stretch when it's unset.

**Review by lishunyang12 — COMMENTED** [review]
The STG path adds a spurious `+cond` when you clear the blocks but leave `stg_scale=1.0` — worth checking that combo.

**Comment by TaffyOfficial** [comment]
> This PR is too large, it won’t continue moving forward. I’ll provide the implementation in smaller, split-out PRs instead.

Given the extensive scope of code modifications, I recommend separating the changes into four independent PRs as follows:
Dedicated PR for I2V relevant changes
Separate PR for 2-stage deployment logic
Independent PR to add new CFG support
Individual PR for official accuracy alignment

File: vllm_omni/diffusion/models/ltx2_3/ltx2_3_guidance.py (hunk 2 above)
Comment (mglyn) [reply vllm_omni/diffusion/models/ltx2_3/ltx2_3_guidance.py:101]: Aappreciate the concerns raised in the review, all of which are valid and will be addressed in the subsequent implementation.
